### DISCUSSION TITLE
Tokenization discussions

### DISCUSSION THREAD

**Discussion by linas** [opencog/link-grammar#1311] [Ideas]
The discussion in #1310 wanders off into tokenization, and a distinct thread for that seems best.  Specifically this comment from @akolonin 
* https://github.com/opencog/link-grammar/discussions/1310#discussioncomment-3052861

I just read the following:
* Jesse O. Wrenn, Peter D. Stetson, Stephen B. Johnson
  An Unsupervised Machine Learning Approach to Segmentation of Clinician-Entered Free Text
  AMIA Annu Symp Proc. 2007; 2007: 811–815.
  PMC2655800
  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2655800/

I found it frustratingly opaque and imprecise.  Work items and clarifications suggest themselves:
 * A precise mathematical, symbolic definition of peak freedom is needed. The
    informal definition is imprecise, prone to misunderstanding, and obscures
    relationships to other similar mathematical formulas.
  * Peak freedom appears to be defined as a second difference, as the
    difference between the prior and the subsequent transition freedom.
    What if, instead, one worked with log2 of transition freedoms, so
    that a peak entropic freedom was a ratio instead of a difference?
  * Given this variant, what other variants of freedom are possible, and
    how are they related to more traditional entropic definitions?
  * Show distribution of peak freedom. That is, there are hundreds of
    thousands of transitions; what's the freedom of each? What's the peak
    freedom of each?
 * How do things change, if one considers not just the next character, but
    the next pair of characters, or the next triple of characters?

@akolonin by "distribution" I mean this: consider taking the bar charts on page 5-6 of your paper (the https://arxiv.org/abs/2205.11443 one) and projecting them onto the right axis. This gives a distribution that shows how often any given height occurs.  The questions: what's that distribution? Is it gaussian? Something else? What's the mean and deviation? ... Personally, I cannot get comfortable with any of this stuff until I see the distributions, and also the formal definitions, and the variants that use log2 of transition freedoms. 

The Wrenn etal 2007 paper and the Kolonin 2022 paper seem to work with "transition freedoms" by exhaustively counting all possible transitions between substrings and single characters.  What if instead, we worked with transitions between substrings and substrings? There are explosively many of these, and so perhaps a random subsampling is more computationally feasible?  This is why I mentioned the "amy" language in https://github.com/opencog/link-grammar/discussions/1310#discussioncomment-3053272 : it provides a random sampling of transitions.

**Comment by linas** [comment]
@akolonin I tried to define transition freedom and peak freedom in a formal fashion, in the attached PDF. Is this correct? Hmm seem I cannot attach, I'm getting github error messages. The PDF is here:  [peak freedom PDF](https://github.com/opencog/learn/blob/master/learn-lang-diary/learn-lang-diary-a1.pdf) please comment.

**Comment by akolonin** [comment]
Hi Linas, I agree that the paper by Wrenn et.al. is pretty vague, speaking softly ;-)
So I was rather guessing what they are doing - see how I compute these metrics myself:
https://github.com/aigents/pygents/blob/main/pygents/token.py#L353

Notice, that the "peak" works good only for Chinese, the "variance" is better for English and Russian.

**Reply by linas** [reply]
I'm trying to puzzle this out. I see this:
```
import pandas as pd
 df = pd.DataFrame(profile_freedoms(model,text ...columns=['pos','gram','f+','f-'])
 df['df+'] = df['f+'].diff() 
 df['peak+'] = df['df+'] - df['df+'].shift(-1)
```

I guess that `df[f+]` is a count of some kind.  Is it a count of the pair `(w,t)` of an n-gram `w` followed by a single-letter token `t` ? Is it some other kind of count?  

What is `df['f+'].diff()` ?  What does the `shift(-1)` do ?

It's important to develop a common math-formula notation for this stuff, that matches what the code does.  It's too much to ask for readers of articles to puzzle out some python code. I think having actual formulas in front of you will clarify "what the heck is going on". ***Please please please*** try to reconcile what is in the PDF with what is in the code.  I'll fix the PDF as needed.

>  "peak" works good only for Chinese, the "variance" is better for English and Russian.

I just updated the [PDF for peak freedom](https://github.com/opencog/learn/blob/master/learn-lang-diary/learn-lang-diary-a1.pdf) to include a kind of log-freedom quantity.  Perhaps they'd work better?

I've done other, unrelated experiments comparing several kinds of jaccard distances (which are a kind of "freedom") with several kinds of different MI formulations, as well as cosine.  I find that cosine is always the worst, in every case, while the others are close to each other, having fairly similar results. Which one wins, and by how much, seems to depend on minor external factors.  You might be seeing something similar.

The idea is that occasional incorrect tokenizations don't matter, because these can be disambiguated later, by higher-order layers. (Bonus if you can assign some probability to each alternative, some "probably good" vs. "not so good" number)  Note that the tokenizer in LG generates alternatives; this is called "the wordgraph", and it's just a collection of "parallel" strings that are alternative tokenizations of the same sentence.  For example:
```
linkparser> we were at the seasand beach
Found 56 linkages (56 had no P.P. violations)
	Linkage 1, cost vector = (UNUSED=0 DIS= 0.00 LEN=10)

    +---->WV---->+     +---------Ju---------+
    +->Wd--+-Spx-+--Pp-+  +--Dmc-+<--SJlp<--+->SJrs->+
    |      |     |     |  |      |          |        |
LEFT-WALL we were.v-d at the seas[&].n and[&].j-n beach.n


Press RETURN for the next linkage.
linkparser> 
	Linkage 2, cost vector = (UNUSED=0 DIS= 0.00 LEN=10)

                       +----------Js----------+
    +---->WV---->+     |  +-------Ds**x-------+
    +->Wd--+-Spx-+--Pp-+  |        +-----A----+
    |      |     |     |  |        |          |
LEFT-WALL we were.v-d at the seasonal[~].a beach.n

Press RETURN for the next linkage.
linkparser> 
	Linkage 7, cost vector = (UNUSED=0 DIS= 0.10 LEN=14)

                       +--------------Js-------------+
                       |  +-----------Ds**x----------+
    +---->WV---->+     |  |      +---------AN--------+
    +->Wd--+-Spx-+--Pp-+  |      |         +----AN---+
    |      |     |     |  |      |         |         |
LEFT-WALL we were.v-d at the sea[&].s sand[&].n-u beach.n
```

Note how the three different tokenizations have a different number of words in the sentence.  This all works automatically.  The "cost" is just the likelihood of it being a correct parse; it there is a cost for the tokenization, it could be added in. Costs are minus log2 of the probability.

**Reply by nickm197** [reply]
In so many unrelated cases log likeliehood ratio is the best function to optimize. I am a bit swamped to repeat Anton's experiments with that.

The Dirac delta formulation of the freedom and the relation to the  power norm is interesting. Your S does not look like an entropy to me, your H does, but that still requires more diligent formulation. One such formulation may be through defining a statistical language model.

**Reply by linas** [reply]
Hi @nickm197 thanks for posting. My goal of writing [that PDF](https://github.com/opencog/learn/blob/master/learn-lang-diary/learn-lang-diary-a1.pdf) is to figure out exactly what it is that Anton has or has not done. 

I won't argue log-liklihoods, entropies and so on, because that is a long and complicated discussion. I'll happily have that argument, just that ... its complicated.

About repeating experiments ... I don't understand the context of that remark.  Is someone asking you to repeat them? I might, maybe, do some experiments similar to these, maybe in 3-6 months. Maybe.  Tokenization is important. But I have a dozen other important tasks.

Regarding random tokenization, the "amy" language in LG has been fixed.  I'm not asking anyone to use it, rather, I'm commenting that maybe its interesting to you.   Here are some examples of words split into two, in random locations, with a random parse tree used to connect up the parts.  Equals signs show the split, and the square brackets mark up the rule used to split:

Bhojpuri ई एगो लमहर वाक्य के परीक्षा ह 
```
    +---------------------------ANY--------------------------+
    |                         +--------------ANY-------------+
    |           +-----ANY-----+              +-------LL------+------
    |           |             |              |               |
LEFT-WALL ई[!ANY-WORD] एगो[!ANY-WORD] लम[!MOR-STEM].= =हर[!MOR-SUFF] 
```
Georgian ეს უფრო გრძელი წინადადების გამოცდაა
```
    +-------------------ANY-------------------+--------------ANY--------------+-------
    |            +------LL-----+------ANY-----+               +-------LL------+
    |            |             |              |               |               |
LEFT-WALL ე[!MOR-STEM].= =ს[!MOR-SUFF] უფრო[!ANY-WORD] გ[!MOR-STEM].= =რძელი[!MOR-SUFF]
```

The bugs that Amir fixed in the last few days had to do with graphemes and unicode code points being different: a "grapheme" is a written unit, which might consist of several unicode codepoints.  Splitting a grapheme into two is fundamentally incorrect, you get garbage.  Only some languages have non-trivial graphemes, see bug report https://github.com/opencog/link-grammar/issues/1315

**Reply by akolonin** [reply]
Hi @linas and @nickm197 - I am swamped till July 26 but I would love to push this work forward after that and submit the results to https://aaai.org/Conferences/AAAI-23/aaai23call/ 
I can also share some latest results on automated finding of hyper-parameters - see slides 40-42:
https://github.com/aigents/pygents/blob/main/docs/2022/experiential-sequential-2022.pdf
and notebooks:
https://github.com/aigents/pygents/blob/main/notebooks/nlp/russian/ru_tokenizer_auto.ipynb
https://github.com/aigents/pygents/blob/main/notebooks/nlp/english/en_tokenizer_auto_brown_compress.ipynb
https://github.com/aigents/pygents/blob/main/notebooks/nlp/chinese/zh_tokenizer_auto.ipynb
@linas I can answer all your outstanding questions on the Zoom call, let me know if you can make it any day 11AM your time or join the group https://t.me/internlp on Telegram to arrange the time.
Cheers!

**Reply by nickm197** [reply]
@linas Thank you for your reply. I have been discussing this work with Anton repeatedly at the early stage, thus the repro comment. 

As for your PDF, I believe that differential formulations are incorrect, because d/dp (N^p) = p*N^(p-1). I may have misunderstood your notation though. To get a logarithm you have to integrate, but that's tricky.

**Reply by linas** [reply]
Careful. It is true that
` d/dN (N^p) = p*N^(p-1)`
however, that is not what is written.  Instead, its the other way around:
```
d/dp (N^p) = d/dp e^(p log N)
         = log N e^(p log N)
         = log N N^p
```

**Reply by nickm197** [reply]
Indeed, you are correct. Sorry.

**Reply by nickm197** [reply]
I have assembled some more thoughts https://github.com/nickm197/INLP/blob/main/heaviside%20and%20entropy.pdf

**Reply by linas** [reply]
Hi Nick. I looked a the PDF. I don't understand the comment about the Heaviside function $H(x)$ it requires a total order on $x$, e.g. that $x$ is a number. But there is no total order on pairs $(w,t)$ -- they're just a bag, an unordered set, just a collection of pairs. The $\Delta$ is just a subset function: for all x in some universe, the x belongs to the subset if $\Delta(x)=1$ and otherwise $\Delta(x)=0$ -- that is why it has the name "indicator function". 

Let me try out some script-fu. Given a set $S\subset U$ the indicator function is

$$\Delta_S(x) = \begin{cases} 1 \text{ if } x\in S \\
 0 \text{ if } x \notin S \end{cases} $$

and vice-versa:

$$ S = \left (  x | x\in U \text{ and }  \Delta(x)=1  \right ) $$

The round parens should have been set-notation curly braces in the above, but apparently that does not display correctly.

**Reply by linas** [reply]
About the norms, yes, you are right, the Banach norm is a real norm; and it's power is not. However, when I take the derivative, I get a different expression that you; it's ugly and it's complicated, and that is why I did not mention it. 

In general, in machine learning, the question of when is something a norm, and when isn't it, what kind of a space is it a norm over, should it be information-theoretic instead, and does it even matter? -- all this is very confusing. I have a long list of "interesting" norm-like, info-theory-like expressions, in several PDF's of my online diaries/research reports (in assorted git repos)

Given Anton's results, I'd like to propose the following: perhaps quasi-norms should be "randomly" generated, and then each can be tried out to see which ones give the best results. There's existing code in MOSES that can generate random expression trees. I was planning on reviving that code anyway, for use in image and audio processing.  It could be used here, too, to manage a pool of "pretty good norms", which are then genetically mutated and crossed-over to create new norms. (This is "genetic programming" or "genetic program evolution", MOSES does genetic programming).

The hard part in the above proposal is to answer the question "what is a good toekenization, and how do I know one when I see it?" Of course, it can be compared to a "gold standard", but is there a way to find out, without a gold standard?

As to physics: the proper procedure is to run the experiment first, find the best (pseudo/quasi)-norm (or the general class of such best norms), and only then turn to theory, and ask "what is the physical interpretation of this?"  In physics, the theory and experiment must go hand in hand.  Right now, the experiments are still very murky and unclear, so it's hard to see what is actually happening.

**Reply by nickm197** [reply]
About delta and Heaviside. I meant that $\Delta{(w, t)} = H(N(w, t)-1/2)$, where $N$ is the count you have introduced and, obviously, maps $(w, t) -> N$. This allows for further differentiable approximations.

**Reply by nickm197** [reply]
I may have lost my ability to differentiate functions over dozens of years not using this ability that was pretty good around 1989. My previous comment on differentiation is an indication. Please feel free to compare notes. The term $\log⁡\Sigma_t(N^p (w,t))$ is intriguing anyway, and all this requires physical explanation.

The relationship of experiment and theory in physics is rather complicated. Basically, what you see in experiment depends heavily on a theory/hypothesis you use, and the difference between experiment and experience is that the first has to rely on some theory/try some defined hypothesis. If you occasionally read Russian and interested n the topic I can refer you to some good reading on that (unfortunately, I don't know English literature on the subject well enough).

I am not opposed to genetic search of a pseudonorm, but the search space is rather large and a norm found this way does not add to understanding the subject which is the goal for me personally.

**Reply by linas** [reply]
Oh! Ahh! Yes, that is actually the classic "fuzzy logic" generalization of set theory. And this is perhaps why things get "fuzzy" -- a strict step function gives you the Jaccard distances, but if you read the wikipedia article on Jaccard, you'll find several "probabilistic Jaccard"  functions, including one that is entropy maximizing (or something like that). If you replace the step function by some sigmoid, then you can start slipping down the path of neural nets.  Or you can insist that you are a right-proper Bayesianst, or an information theorist, and just work with probabilities. This is partly why I proposed "random" metrics.

I'm trying to learn Russian, very slowly, so,  правоверный Байесианец  I guess

**Reply by linas** [reply]
I'm half-joking about правоверный. You're presumably interested in AGI. You have to be very careful with beliefs. Once you start believing in something, it can blind you to the actual truth. Beliefs prevent realities from being perceived. It's like the drunk searching for his keys under a street lamp. The policeman asks "where did you lose them?" "Over there, but there's more light over here."  This is why I want to run the experiments, to search for the answer: instead of just declaring that the solution is known and it is X.  But it is difficult: how do you search? How do you know if you've found the answer? This is the metaphysics...  perhaps this sounds simplistic, but I struggle with this daily.

**Reply by linas** [reply]
> the search space is rather large

Yes, it is, but the search is automated: you just let the computer run for a few days or weeks. Perhaps it will discover a very classic formula from information theory, and perhaps it will discover something else. "Understanding" is like trying to hold water with your hands.

> Derivatives

Here, try this:

$$ \begin {align}  \frac{d}{dp} \left \Vert N \right \Vert_p &= \frac{d}{dp} \left [ \sum N^p  \right ]^{1/p} \\
&= \frac{d}{dp} \exp \left(  \frac{1}{p} \log \left[ \sum N^p \right] \right) \\
&= \left \Vert N \right \Vert_p  \frac{d}{dp} \left(  \frac{1}{p} \log \left[ \sum N^p \right] \right) \\
&= \cdots 
\end{align}
$$

and if you finish that, you get a longer, complicated mess.  Perhaps it has "physical meaning", and perhaps it is just a mess. Hard to say.

**Reply by linas** [reply]
If I didn't make any mistakes, then for the $p=1$ case I get

$$  \left . \frac{d}{dp} \left \Vert N \right \Vert_p \right |_{p=1} = \sum N\log N -   \left \Vert N \right \Vert_1 \log  \left \Vert N \right \Vert_1 $$

Note that if the count is normalized, so that $\left \Vert N \right \Vert_1 = 1$ then the second term is zero.

**Comment by akolonin** [comment]
@nickm197 @linas - we have extended the INLP workshop deadline https://aigents.github.io/inlp/2022/ 
Do you think this topic worth discussing there so I make submission on my latest results and you @nickm197 submit your latest considerations?

**Reply by nickm197** [reply]
@akolonin I see the current discussion as an introductory on a wide array of topics. I would expect it would branch into several more specific ones. As for the tokenization, this specific discussion seems to be rather early,  although promising. As for my considerations, the most finalized relevant considerations are the ones on the context and criticism of Yuret dissertation. Please let me know what do you think worth submitting.

**Reply by linas** [reply]
I too got swamped ... Apologies for the late reply. Perhaps we could have a discussion panel  reviewing these topics, with a goal of developing a strategy or plan to move forward?

**Comment by nickm197** [comment]
> that is actually the classic "fuzzy logic" generalization of set theory

Indeed, one can look at it this way

>  You're presumably interested in **AGI**

Who is not these days?

> Once you start believing in something, it can blind you to the actual truth. Beliefs prevent realities from being perceived.

This setting is the exact one where the classical Popperian discourse is valid (and it is valid until the knowledge field becomes so entangled that finally you can only falsify - in popperian terms - the whole mentis): You work with hypotheses and keep them falsifiable. In other words, there is no actual truth, but a succession of models that increasingly better describe reality. A model is an explanation of reality that accepts a certain viewpoint of the reality and, thus, compresses information about reality with losses. Occam's razor is a requirement to improve compression of information in a model.  

> Perhaps it will discover a very classic formula from information theory, and perhaps it will discover something else. 

I am in no way an expert in genetic algorithms, but always considered them to be a descent in a rather fancy discrete space. Thus, the result depends on the way you construct this space more then on anything else. In math, one would need to prove that the optimal solution belongs to that space.

**Reply by linas** [reply]
A week ago, I prepared a long reply to the above as I prepared to go to sleep. Then matters intervened.

Some commentary on genetic algorithms, search spaces, and refining models that are ever-more accurate representations of reality.  Consider the conventional decision tree: if X happens, you should do Y, unless of course Z and then you should do W (and still do Y sometimes, whenever Q is there..) This is a branching tree of  decisions, and if you want to make it match "reality", it gets ever more complex. 

The complexity of the tree is combinatorially exploding.  If you have only two input variables, you can list all possibilities on a small piece of paper. Once you have more than 5 or 6, this becomes impractical. More than a few dozen, and even good software grinds to a halt. The space becomes much too large to explore exhaustively; there are too many possibilities to try.

The genetic-programming approach to solving the above is to find some small, simple trees that approximately model the situation, and then either make some small changes that improve accuracy (mutation) or combine two models to get a larger, more complex model  (cross-over) 

If you are training with a well-specified objective function, then, yes, it is a form of descent: one is looking to lower the score with each attempt.  As to optimality: You can "prove" that there is an "optimal solution"; the hard part is actually finding it. This is the generic problem in machine learning.

------
Anyway, the proposal that I am making is to use genetic programming to explore the space of different possible text segmentation algorithms.  The only benefit of using genetic programming is that it provides an existing software framework for managing a collection of trees, and for mutating them in a controlled way. So, instead of trying this and that segmentation algo by hand, the idea is to automate the search.

-----
I can provide more detail on how genetic programming can be made to work, and what some of the traps and pitfalls are, if you are interested. My general goal is to build up enough software infrastructure to allow text segmentation to be automated, and to build it generally enough that audio and visual data can be segmented too. Or even arbitrary sensory data, e.g. magnetic field readings or whatever.  ... anyway, the goal is to build the software that can confront the general case.

-----
There are other things to talk about, but I'm trying to keep the discussion narrow.

**Reply by nickm197** [reply]
There is quite a wide literature on both unsuperwised text and image segmentation... Then, there are things like phoneme segmentation in speech etc. There are some recurrent topics, but at the moment I would say I lack understanding to generalize, even that I am no stranger to several of these tasks.

Keeping things narrow, I have played with @akolonin old (April?) notebooks a bit. My current perception is that the freedom-based aproach is a fancy way to detect seperators in a text.

Experimental protocol: remove whitespace from a corpus in Russian or English. 

Theoretical ground: if we consider an end of a word, it is more likely to find a separator (say, 
{' ','-','.',',','?','"',':',';','&','(',')','!','\n',' '})
 after it, rather then find the same n-gram inside a different word. The effect of finding separators deteriorates as the corpus becomes larger and thus includes more rare words and typos. In a limit of an infinite corpus the freedom of all n-grams would become equal and equal to the size of alphabet. 

It looks like it is possible to find separators explicitly in an unsuperwised manner. This can be approached by finding symbols that are most independent from their surrounding = minimum mutual information with the neighbours or minimum log likelihood ratio. But, unlike what Anton already did, this can be done across the corpus, not per locus. What do you @akolonin @linas think?

The above, if works, also does not solve the segmentation problem for writings without separators, like Chinese. To that end I was thinking about the definition of a word: "a speech sound or series of speech sounds that symbolizes and communicates a meaning usually without being divisible into smaller units capable of independent use". While the semantic part is hard to grasp computationally, the indivisability potentially can be. For example, I was thinking about the LZ78-type algorithm with some kind of corpus-wide optimization, but didn't quite come up with the target function (yet).

**Comment by akolonin** [comment]
Hi @linas can you email me your slides for your talk on August 19? 
https://aigents.github.io/inlp/2022/
How long will it take? 
Will 45 minutes for the talk and 14 minutes for discussion suit you?

**Reply by linas** [reply]
I haven't prepared the slide yet; i expect them to be similar to those of last year. yes, 45+14 is fine.

**Comment by akolonin** [comment]
@nickm197 we have got potentially relevant reference to our work https://www.researchgate.net/publication/361505964_Estimating_Sentence-like_Structure_in_Synthetic_Languages_Using_Information_Topology/references

**Reply by nickm197** [reply]
Interesting, thank you - I'll take a look

**Comment by linas** [comment]
Hi Nick, I want to continue the conversation but have been distracted by other matters. Are you in (or would you be able to get to) Europe next month?

**Comment by akolonin** [comment]
@linas @nickm197 we will meet on Aug 19, check your email and workshop agenda - need your slides by then ;-) https://aigents.github.io/inlp/2022/

**Comment by akolonin** [comment]
@linas - you are about to present on Aug 19 at https://aigents.github.io/inlp/2022/ and on Aug 22 at http://agi-conf.org/2022/schedule/  
We need your slides and yourself to to the test Zoom call before. Please check your email and reply.
